README.md
# epidatR

This package reads [EpiData](https://www.epidata.dk/) epx-files directly into R. So far it contains one major function `read.EpiData()`. The other functions in this repository are helpers.

Compared to the allready existing package [epxToR](https://cran.r-project.org/package=epxToR), `read.EpiData()` not only transfers epx-files into an R-`data.frame`. Additionally, values defined as missing in EpiData will be set to `NA`, defined value and variable labels will be applied and variables will be converted to appropriate R-classes according to their field type in EpiData. 

If this conversion is not wanted or causes some trouble, the original raw codes can be transferred into the `data.frame` as `character`-variables, by setting the argument `convert="none"`.

If the epx-file consists of one data set only, the output will be a `data.frame`. If the epx-file contains a relational data base with multiple data sets, it will be a `list` of multiple `data.frame`s.

Variable labels, study information and information about the relation of data sets are given as attributes.

Helper functions are: 

* `epx.extract()` extracts the xml-like data from the epx-file.
* `epx.read()` transfers the extracted data into a `data.frame`.
* `epx.missing()` sets defined missings to `NA`.
* `epx.class()` applies value labels and variable classes.
* `epx.example()` returns the path of the example files in the attached package.


The function works fine with the example files "SomeFakeData.epx", "marathon.epx", "sample.v3.epx", "Clinical_Example.epx" and "OneLine.epx".

The function can not jet read encrypted epx-files. I would need much more expertise in cryptography to implement this feature. Perhaps there are some experts to help me out. For experimentation I have added the files "sample.v3_password.epx" with simple (symetric) encryption (Password = "password") and "sample.v3_ExtendedAccess.epx" with asymetric encryption produeced by the Extended Access module (Login = "admin", Password = "password"). 

The package is not jet available from CRAN. You can install it from here, using the devtools-package.

```{r}
devtools::install_github("https://github.com/JohannPopp/epidatR")
```

## Still To Do List
| Date       | To do                   | Solved | Notes                       |
|------------|-------------------------|--------|-----------------------------|
| 2023-11-28 | Reading encrypted files |        | complicated                 |
| 2023-11-28 | Setting the time zone   | 2023-11-28 | generally set to "UTC"  |
| 2023-11-28 | Writing epx files       |        | lot of work. Do we need it? |


 <!-- badges: start -->
  [![R-CMD-check](https://github.com/JohannPopp/epidatR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/JohannPopp/epidatR/actions/workflows/R-CMD-check.yaml)
  <!-- badges: end -->




NEWS.md
# epidatR 0.3.0

* Initial correctly working GitHub submission. 
* Reads simple and relational data sets.
* Converts defined missing values to NA (by default).
* Replaces value codes with value labels (by default).
* Applies object classes according to the field types and label sets defined in the EpiData file (by default).

# epidatR 0.3.1

* Adds the variable "status" for the record status defined in EpiData.
* Restructures the Tests so that they do not depend on the package *digest* any more.

# epidatR 0.3.2

* Time zone is set to "UTC" for all date and time classes.


tests/testthat.R
library(testthat)
library(epidatR)

test_check("epidatR")




tests/testthat/test-labels.R
# Test replacement of value codes with value labels
test_that("Labels-SomeFake", {
  expect_equal(paste(
    unlist(
      epx.labels(
        epx.read(epx.extract(epx.example("SomeFakeData.epx"))$perDataSet$ds1),
        epx.extract(epx.example("SomeFakeData.epx"))$perDataSet$ds1)),
    collapse = ", "),
    "1, 2, 3, 4, 03/11/2023, 10/11/2023, 30/10/2023, NA, 65, no data, 18, 39, male, no data, female, male, 189, 160, no data, 170, 78, 52, 60, 81, No notes, NA, NA, This is fake data")
})


tests/testthat/test-examples.R
test_that("Clinical example", {
  expect_equal(digest::digest(read.EpiData(epx.example("Clinical_Example.epx")), "sha256"),
               "dbe4b3e74bb408ebe0c6be106a7d7d296b8a11c11003b8ddaf92818871f16074")
})


test_that("One line dataframes", {
  expect_equal(digest::digest(read.EpiData(epx.example("OneLine.epx")), "sha256"),
               "109f73215998a2206395b116863e8b6044ad8077e2e25ead5030c99f47dfcdd1")
})

test_that("sample.v3.epx", {
  expect_equal(digest::digest(read.EpiData(epx.example("sample.v3.epx")), "sha256"),
               "b28f1ecb88cd4619c3cfcdb44eab7f6d78eaf605578162da1b2ee5d79ef4c3c9")
})

test_that("Marathon", {
  expect_equal(digest::digest(read.EpiData(epx.example("marathon.epx")), "sha256"),
               "6a9f5333c3eb91a96793ce9ebccea66a87c730b76fa3e935cb5c64e792bcf1e4")
})


R/setLabels.R
# Function to replace value codes with value labels extracted from an epx-file.


# This is part of a project to create a package to read epx-files from EpiData into R

# Johann Popp
# 2023-11-20
# Last modification 2023-11-22
###########################################


#' Helper: Replace value codes with value labels extracted from an epx-file
#'
#' A helper function of \code{\link{read.EpiData}}, to replace values by value labels and convert variables to appropriate classes.
#'
#' @param dat A data.frame or a list of data.frames created by \code{\link{epx.extract}}.
#' @param info Additional information about the EpiData-file created by \code{\link{epx.extract}}.
#'
#' @return A data.frame containing value labels instead of value codes.
#' @export
#'
#' @examples
#' # This demonstrates the effects of all six helper functions to \code{\link{read.EpiData}}.
#'
#' (x <- epx.example("SomeFakeData.epx"))
#'
#' # epx.extract() extracts the information from an EpiData-epx file
#' (info <- epx.extract(x))
#'
#' # epx.read() transforms the information into a data.frame
#' (dat <- lapply(info$perDataSet, epx.read))

#' # Combine dat and info in a data-set-wise list
#' # This is a necessary detour to handle EpiData files with multiple data sets.
#' perDataSet <- mapply(function(dat, info) list(list(dat = dat, info = info)), dat, info[[7]])
#'
#' # epx.labels() replaces value codes with value labels
#' (datLab <- epx.labels(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.missing() replaces definde missing values with NA
#' (datMis <- epx.missing(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.class() sets variable classes according to the field types defined in EpiData
#' (datClass <- epx.class(perDataSet$ds1$dat, perDataSet$ds$info))
#'
#' # Bringing it all together
#' read.EpiData(epx.example("SomeFakeData.epx"))



epx.labels <- function(dat, info){


  # Extract field types
  fieldTypes <- info$fieldTypes
  # Extract value lable sets that contain codes for non-missing values
  catValLabelSets <- info$valLabelSets#[lapply(info$valLabels, function(x) sum(!grepl("missing=\"true\"", x))) > 0]
  # Extract value labels
  catValLabels <- lapply(catValLabelSets, xml2::xml_children)
  # value codes
  valCodes <- lapply(catValLabels, function(x) xml2::xml_attr(x, "value"))
  # Value labels
  validLabels <- lapply(catValLabels, xml2::xml_text)


  # Indicate which value set matches which variable
  indexValLabelSet  <- as.numeric(
    lapply(
      info$fieldValLabSets,
      function(x){
        which(
          grepl(
            paste0("^", x, "$"),
            xml2::xml_attr(catValLabelSets, "id")))
      }
    )
  )


  # Replace value codes with value labels
  dat2 <-
    mapply(
      function(x, y, z){
        if(!is.na(y) & (z == "ftInteger" | z == "ftFloat")){

          validLabels[[y]][as.numeric(lapply(x, function(x)  which(x == valCodes[[y]])))]
        } else {
          if(!is.na(y) & z == "ftString"){
            validLabels[[y]][factor(x, levels = valCodes[[y]])]
          } else {x}
        }
      },
      x = dat, y = indexValLabelSet, z = fieldTypes)
  dat2[is.na(dat2)] <- dat[is.na(dat2)]


  ifelse(nrow(dat) >1,
         dat <- as.data.frame(dat2, stringsAsFactors = FALSE),
         dat <- as.data.frame(t(dat2), stringsAsFactors = FALSE))




  # Reassign variable labels
  attributes(dat)$variable.labels <- info$fieldLabels



  dat
}




R/missingValues.R
# Function to convert codes, that are defines as missing values into NA in the data.frame
# This is part of a project to create a package to read epx-files from EpiData into R

# Johann Popp
# 2019-06-09
# Last modification: 2023-11-14
###########################################


#' Helper: Convert defined missings to NA
#'
#' A helper function of \code{\link{read.EpiData}}, to convert defined missing values to NA.
#'
#' @param dat A data.frame or a list of data.frames created by \code{\link{epx.extract}}.
#' @param info Additional information about the EpiData-file created by \code{\link{epx.extract}}.
#'
#' @return A data.frame with defined missing values set to NA.
#' @export
#'
#' @examples
#' # This demonstrates the effects of all six helper functions to \code{\link{read.EpiData}}.
#'
#' (x <- epx.example("SomeFakeData.epx"))
#'
#' # epx.extract() extracts the information from an EpiData-epx file
#' (info <- epx.extract(x))
#'
#' # epx.read() transforms the information into a data.frame
#' (dat <- lapply(info$perDataSet, epx.read))
#'
#' # Combine dat and info in a data-set-wise list
#' # This is a necessary detour to handle EpiData files with multiple data sets.
#' perDataSet <- mapply(function(dat, info) list(list(dat = dat, info = info)), dat, info[[7]])
#'
#' # epx.labels() replaces value codes with value labels
#' (datLab <- epx.labels(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.missing() replaces definde missing values with NA
#' (datMis <- epx.missing(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.class() sets variable classes according to the field types defined in EpiData
#' (datClass <- epx.class(perDataSet$ds1$dat, perDataSet$ds$info))
#'
#' # Bringing it all together
#' read.EpiData(epx.example("SomeFakeData.epx"))

epx.missing <- function(dat, info){
  dat <- dat
  info <- info

  # Extract value codes
  valueCodes <- lapply(info$valLabels, function(x) xml2::xml_attr(x, "value"))
  # Identify value label sets that have codes for missing values and where are those codes
  indexMissing <- lapply(info$valLabels, function(x) grep("missing=\"true\"", x))
  # Identify missing value codes
  missingCodes <- as.list(mapply(function(x, y) x[y], valueCodes, indexMissing))

  # Index the corresponding value label set for each of the variables
  indexValLabSet <- lapply(info$fieldValLabSets, function(x) which(xml2::xml_attr(info$valLabelSets, "id") == x))
  # Missings per variable
  missingsPerVar <- lapply(missingCodes[as.numeric(paste(indexValLabSet))],
                           paste, collapse = "|")
  missingsPerVar[missingsPerVar == ""] <- NA

  # # Convert defined missing values into NA
  # dat[
  #   mapply(function(x, y){
  #     grepl(y, x)
  #   },
  #   x = dat,
  #   y = missingsPerVar) == 1
  #   ] <- NA

  # Identify defined missing values in the data.frame
  defMissings <-
    mapply(function(x, y){
      grepl(y, x)
    },
    x = dat,
    y = missingsPerVar)

  # Convert defined missing values to NA
  ifelse(nrow(dat) >1,
         dat[defMissings] <- NA,
         dat[,which(defMissings)] <- NA)

  # Convert undefined missing values in text fields (expressed as "NA") into NA
  dat[dat == "NA"] <- NA

  dat
}


R/convertToClasses.R
# Function to convert into R-object classes according to the field types of the epx-data.

## So far the function can read and convert epx.Projects with one data set and without encryption.

# This is part of a project to create a package to read epx-files from EpiData into R

# Johann Popp
# 2019-06-11
# Last modification 2023-11-28
###########################################


#' Helper: Show value labels and convert variables to appropriate classes
#'
#' A helper function of \code{\link{read.EpiData}}, to replace values by value labels and convert variables to appropriate classes.
#'
#' @param dat A data.frame or a list of data.frames created by \code{\link{epx.extract}}.
#' @param info Additional information about the EpiData-file created by \code{\link{epx.extract}}.
#'
#' @details
#' All date and time fields are convertet using *Coordinated Usual Time (UTC)* as time zone argument as EpiData does not provide time zone information by default.
#'
#'
#' @return A data.frame with value labels and appropriate object classes of the variables.
#' @export
#'
#' @examples
#' # This demonstrates the effects of all six helper functions to \code{\link{read.EpiData}}.
#'
#' (x <- epx.example("SomeFakeData.epx"))
#'
#' # epx.extract() extracts the information from an EpiData-epx file
#' (info <- epx.extract(x))
#'
#' # epx.read() transforms the information into a data.frame
#' (dat <- lapply(info$perDataSet, epx.read))
#'
#' # Combine dat and info in a data-set-wise list
#' # This is a necessary detour to handle EpiData files with multiple data sets.
#' perDataSet <- mapply(function(dat, info) list(list(dat = dat, info = info)), dat, info[[7]])
#'
#' # epx.labels() replaces value codes with value labels
#' (datLab <- epx.labels(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.missing() replaces definde missing values with NA
#' (datMis <- epx.missing(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.class() sets variable classes according to the field types defined in EpiData
#' (datClass <- epx.class(perDataSet$ds1$dat, perDataSet$ds$info))
#'
#' # Bringing it all together
#' read.EpiData(epx.example("SomeFakeData.epx"))


epx.class <- function(dat, info){

  # Extract field types
  fieldTypes <- info$fieldTypes
  # # Extract value lable sets that contain codes for non-missing values
  # catValLabelSets <- info$valLabelSets#[lapply(info$valLabels, function(x) sum(!grepl("missing=\"true\"", x))) > 0]
  # Extract value labels
  catValLabels <- lapply(info$valLabelSets, xml2::xml_children)
  # # value codes
  # valCodes <- lapply(catValLabels, function(x) xml2::xml_attr(x, "value"))
  # # Value labels
  # validLabels <- lapply(validCategories, xml2::xml_text)
  # Names of value label sets
  valLabSets <- xml2::xml_attr(info$valLabelSets, "id")



  # Indicate which value set matches which variable
  indexValLabelSet  <- as.numeric(
    lapply(
      info$fieldValLabSets,
      function(x){
        which(
          grepl(
            paste0("^", x, "$"),
            xml2::xml_attr(info$valLabelSets, "id")))
      }
    )
  )

  # Which label sets contain only labels for missing values?
  missOnlySet <-
    valLabSets[
      unlist(
        lapply(catValLabels, function(x){
          all(grepl("missing=\"true\"", x))
        })
      )]


    # Convert integer fields
    toInt <- (grepl(paste0("^", missOnlySet, "$", collapse = "|"), info$fieldValLabSets) |
                is.na(info$fieldValLabSets)) &
      grepl("ftInteger|ftAutoInc", fieldTypes)

    dat[,toInt] <-
      as.integer(
        unlist(
          dat[,toInt])
      )

    # Convert numeric fields
    toNum <- (grepl(paste0("^", missOnlySet, "$", collapse = "|"), info$fieldValLabSets) |
                is.na(info$fieldValLabSets)) &
      grepl("ftFloat", fieldTypes)

    if(sum(toNum) > 1){
      dat[, toNum] <-
        apply(dat[, toNum], 2, function(x) sub(info$infoSeparators[3], ".", x))
      dat[, toNum] <-
        apply(dat[, toNum], 2, as.numeric)
    } else {
      if(sum(toNum) == 1){
        dat[, toNum] <-
          as.numeric(
            sub(
              info$infoSeparators[3],
              ".",
              dat[, toNum]
            ))
      }
    }


    # Convert categorical fields
    notMissOnlySet <-
      valLabSets[!grepl(paste0("^", missOnlySet, "$", collapse = "|"), valLabSets)]

    toFac <- grepl(paste0("^", notMissOnlySet, "$", collapse = "|"), info$fieldValLabSets) |
      fieldTypes == "ftString"

    if(sum(toFac) == 1){
      dat[, toFac] <-
        factor(dat[, toFac])
    } else {
      if(sum(toFac) > 1){
        # dat[, !is.na(indexValLabelSet)] <-
        #   apply(dat[, !is.na(indexValLabelSet)], 2, factor)
        dat[, toFac] <-
          as.data.frame(lapply(dat[, toFac], factor))
      }
    }


    # Convert date fields
    ## DMY format
    dmyPattern <- paste("%d", info$infoSeparators[[1]][1], "%m", info$infoSeparators[[1]][1], "%Y", sep = "")
    if(sum(fieldTypes == "ftDMYDate" | fieldTypes == "ftDMYAuto") > 1){
      dat[, fieldTypes == "ftDMYDate" | fieldTypes == "ftDMYAuto"] <-
        lapply(dat[,fieldTypes == "ftDMYDate" | fieldTypes == "ftDMYAuto"], strptime, dmyPattern, tz = "UTC")
    } else {
      dat[,fieldTypes == "ftDMYDate" | fieldTypes == "ftDMYAuto"] <-
        data.frame(strptime(dat[,fieldTypes == "ftDMYDate" | fieldTypes == "ftDMYAuto"], dmyPattern, tz = "UTC"))
    }

    ## MDY format
    mdyPattern <- paste("%m", info$infoSeparators[[1]][1], "%d", info$infoSeparators[[1]][1], "%Y", sep = "")
    if(sum(fieldTypes == "ftMDYDate" | fieldTypes == "ftMDYAuto") > 1){
      dat[, fieldTypes == "ftMDYDate" | fieldTypes == "ftMDYAuto"] <-
        lapply(dat[,fieldTypes == "ftMDYDate" | fieldTypes == "ftMDYAuto"], strptime, mdyPattern, tz = "UTC")
    } else {
      dat[,fieldTypes == "ftMDYDate" | fieldTypes == "ftMDYAuto"] <-
        data.frame(strptime(dat[,fieldTypes == "ftMDYDate" | fieldTypes == "ftMDYAuto"], mdyPattern, tz = "UTC"))
    }

    ## YMD format
    ymdPattern <- paste("%Y", info$infoSeparators[[1]][1], "%m", info$infoSeparators[[1]][1], "%d", sep = "")
    if(sum(fieldTypes == "ftYMDDate" | fieldTypes == "ftYMDAuto") > 1){
      dat[, fieldTypes == "ftYMDDate" | fieldTypes == "ftYMDAuto"] <-
        lapply(dat[,fieldTypes == "ftYMDDate" | fieldTypes == "ftYMDAuto"], strptime, ymdPattern, tz = "UTC")
    } else {
      dat[,fieldTypes == "ftYMDDate" | fieldTypes == "ftYMDAuto"] <-
        data.frame(strptime(dat[,fieldTypes == "ftYMDDate" | fieldTypes == "ftYMDAuto"], ymdPattern, tz = "UTC"))
    }

    # Convert time
    timePattern <- paste("%H", info$infoSeparators[2], "%M", info$infoSeparators[2], "%S", sep = "")
    if(sum(fieldTypes == "ftTime" | fieldTypes == "ftTimeAuto") > 1){
      dat[, fieldTypes == "ftTime" | fieldTypes == "ftTimeAuto"] <-
        lapply(dat[,fieldTypes == "ftTime" | fieldTypes == "ftTimeAuto"], strptime, timePattern, tz = "UTC")
    } else {
      dat[,fieldTypes == "ftTime" | fieldTypes == "ftTimeAuto"] <-
        data.frame(strptime(dat[,fieldTypes == "ftTime" | fieldTypes == "ftTimeAuto"], timePattern, tz = "UTC"))
    }

    # Convert to logical fields
    dat[, fieldTypes == "ftBoolean"] <-
      dat[, fieldTypes == "ftBoolean"] == "Y"


  # Reassign variable labels
  attributes(dat)$variable.labels <- info$fieldLabels



  dat
}




R/readIntoDataFrame.R
# Function to convert exracted information from an epx-file into a data.frame
# This is part of a project to create a package to read epx-files from EpiData into R

# Johann Popp
# 2019-06-23
# Last update: 2023-11-27
###########################################


#' Helper: Reading xml-data from a EpiData-epx file into a data.frame.
#'
#' A helper function of \code{\link{read.EpiData}} to read xml-data into a data.frame
#'
#' @param x Data set informations extracted by \code{\link{epx.extract}} (info [[7]]).
#'
#' @return A data.frame
#' @export
#' @examples
#' # This demonstrates the effects of all six helper functions to \code{\link{read.EpiData}}.
#'
#' (x <- epx.example("SomeFakeData.epx"))
#'
#' # epx.extract() extracts the information from an EpiData-epx file
#' (info <- epx.extract(x))
#'
#' # epx.read() transforms the information into a data.frame
#' (dat <- lapply(info$perDataSet, epx.read))
#'
#' # Combine dat and info in a data-set-wise list
#' # This is a necessary detour to handle EpiData files with multiple data sets.
#' perDataSet <- mapply(function(dat, info) list(list(dat = dat, info = info)), dat, info[[7]])
#'
#' # epx.labels() replaces value codes with value labels
#' (datLab <- epx.labels(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.missing() replaces definde missing values with NA
#' (datMis <- epx.missing(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.class() sets variable classes according to the field types defined in EpiData
#' (datClass <- epx.class(perDataSet$ds1$dat, perDataSet$ds$info))
#'
#' # Bringing it all together
#' read.EpiData(epx.example("SomeFakeData.epx"))


epx.read <- function(x){
  if(length(x$records > 0)){
  records <- x$records
  fieldNames <- x$fieldNames
  fieldLabels <- x$fieldLabels
  fieldTypes <- x$fieldTypes

  # Add category "NA" for missing values in the records
  records <- paste("NA;", records, sep="")
  # Construct a pattern to seperate individual variables in the records
  pattern <- paste(";", paste(fieldNames, "=", sep = "", collapse = "|;"), sep = "")

  # Enter values from the records into a data.frame.
    dat <- as.data.frame(
      lapply(
        # colnames(dat),
        fieldNames,
        function(y){
          records2 <- gsub(paste("[[:print:]]*", ";", y, "=", sep = ""), "", records)
          unlist(lapply(strsplit(records2, pattern), function(x) x[1]))
        }),
      stringsAsFactors = FALSE, col.names = fieldNames)

  # Delete double quotation marks at the beginning and end of text fields
  characterField <- grepl("ftString|ftMemo|ftUpperString", fieldTypes)
  if(sum(characterField) == 1) {
    dat[,characterField] <-
    gsub("^\"|\"$", "", dat[,characterField])
  } else {
    # if(sum(grepl(characterPattern, fieldTypes)) > 1) {
    if(sum(characterField) >1){
      # dat[,grepl(characterPattern, fieldTypes)] <-
      dat[,characterField] <-
        apply(dat[,characterField], 2, function(x) gsub("^\"|\"$", "", x))
    }
  }



  # Include variable labels
  attributes(dat)$variable.labels <- fieldLabels


  dat} else {"This data frame has no entries."}
}



R/extractInformation.R
# Function to extract xml-type information from a epx-file that was produced with EpiData 4.6.
# This expands the function to the case of multiple data sets in one file
# This is part of a project to create a package to read epx-files into R

# Johann Popp
# 2019-06-08
# Last update: 2023-11-27
###########################################


#' Helper: Extract Data From Epx-File
#'
#' A helper function of \code{\link{read.EpiData}}, to extract the xml-like data from an epx-file.
#'
#' @param x An EpiData epx-file
#'
#' @return
#'     A list, containing the following elements:
#'     \describe{
#'       \item{$epx:}{The entire xml_document from the epx-file.}
#'       \item{$infoEpiData:}{Information about EpiDatas set up.}
#'       \item{$infoStudy}{Information about the study.}
#'       \item{$infoSeparators}{Used separators.}
#'       \item{$infoDataSets}{Summary information about the data sets.}
#'       \item{$infoParentDataSet}{Information about parent data sets in relational data bases.}
#'       \item{$perDataSet}{Information from each data set.}
#'       \item{$infoKeyVars}{Key variables in relational data bases.}
#'    }

#' @export
#' @examples
#' # This demonstrates the effects of all six helper functions to \code{\link{read.EpiData}}.
#'
#' (x <- epx.example("SomeFakeData.epx"))
#'
#' # epx.extract() extracts the information from an EpiData-epx file
#' (info <- epx.extract(x))
#'
#' # epx.read() transforms the information into a data.frame
#' (dat <- lapply(info$perDataSet, epx.read))

#' # Combine dat and info in a data-set-wise list
#' # This is a necessary detour to handle EpiData files with multiple data sets.
#' perDataSet <- mapply(function(dat, info) list(list(dat = dat, info = info)), dat, info[[7]])
#'
#' # epx.labels() replaces value codes with value labels
#' (datLab <- epx.labels(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.missing() replaces definde missing values with NA
#' (datMis <- epx.missing(perDataSet$ds1$dat, perDataSet$ds1$info))
#'
#' # epx.class() sets variable classes according to the field types defined in EpiData
#' (datClass <- epx.class(perDataSet$ds1$dat, perDataSet$ds$info))
#'
#' # Bringing it all together
#' read.EpiData(epx.example("SomeFakeData.epx"))

epx.extract <- function(x){
  # Read epx-data and remove name spaces.
  epx <- xml2::xml_ns_strip(xml2::read_xml(x))

  # Extract general Informations
  infoEpiData <- xml2::xml_attrs(xml2::xml_find_all(epx, "//EpiData"))
  infoStudy <- xml2::xml_find_all(epx, "//StudyInfo")
  infoSeparators <- unlist(xml2::xml_attrs(xml2::xml_find_all(epx, "//Settings")))

  infoDataSets <- xml2::xml_find_all(epx, "//DataFile")

  # Data set relations
  level <- xml2::xml_find_all(epx, "//DataFileRelations") |>
    xml2::xml_children()
  infoRelations <- data.frame(data.set = xml2::xml_attr(level, "dataFileRef"))
  level <- lapply(level, xml2::xml_find_all, ".//DataFileRelation")
  child <- lapply(level, xml2::xml_attr, "dataFileRef")
  child <- unlist(lapply(child, paste, collapse = "; "))
  infoRelations$offspring.data.sets = child

  # Key variables
  infoKeyVars <- data.frame(
    data.set =
      xml2::xml_find_all(epx, "//DataFile") |>
      xml2::xml_attr("id"),
    labels =
      xml2::xml_find_all(epx, "//DataFile") |>
      xml2::xml_child("Caption") |>
      xml2::as_list() |> unlist(),
    key =
      xml2::xml_find_all(epx, "//DataFile") |>
      lapply(xml2::xml_child, "KeyFields") |>
      lapply(xml2::xml_children) |>
      lapply(xml2::xml_attr, "fieldRef") |>
      lapply(paste, collapse = "; ") |>
      as.character()
  )

  # Merge with infoRelations
  infoRelations <-
    merge(infoRelations, infoKeyVars, by = "data.set")[,c(1,3,2,4)]

  # Extract information for each data set
  epxExtractDataSet <- function(df){

    # Extract data base entries
    records <-
      xml2::xml_text(
        xml2::xml_find_all(
          xml2::xml_find_all(df, "Records"),
          "Record")
      )

    # Extract record status
    recordStatus <-
      substring(
        xml2::xml_attr(
          xml2::xml_children(
            xml2::xml_find_all(df, "Records")
          ),
          "status"
        ),
        3)

    # Extract field definitions
    datFields <- xml2::xml_find_all(df, ".//Field")
    # Field names
    fieldNames <- xml2::xml_attr(datFields, "id")
    # Field labels
    fieldLabels <- xml2::xml_text(datFields)
    # Field types
    fieldTypes <- xml2::xml_attr(datFields, "type")

    # Value label sets
    # Indicate which fields have which value label sets
    fieldValLabSets <- xml2::xml_attr(datFields, "valueLabelRef")
    # Extract value labels (all value label sets for all data sets)
    valLabelSets <- xml2::xml_find_all(epx, "//ValueLabelSet")
    valLabels <- lapply(valLabelSets, function(x) xml2::xml_find_all(x, ".//ValueLabel"))

    # Extract separators
    infoSeparators <- infoSeparators



    # Gather information in a list
    list(records = records, recordStatus = recordStatus, datFields = datFields,
         fieldNames = fieldNames, fieldLabels = fieldLabels,
         fieldTypes = fieldTypes, fieldValLabSets = fieldValLabSets, valLabelSets = valLabelSets,
         valLabels = valLabels, infoSeparators = infoSeparators)

  }

  perDataSet <- lapply(infoDataSets, epxExtractDataSet)
  names(perDataSet) <- xml2::xml_attr(infoDataSets, "id")





  # Gather information in a list
  list(epx = epx, infoEpiData = infoEpiData, infoStudy = infoStudy, infoSeparators = infoSeparators,
       infoDataSets = infoDataSets, infoRelations = infoRelations,
       perDataSet = perDataSet)
}





R/readEpiData.R
# Function to combine the single functions of the project and bring everthing together.

# Arguments:
# x: an epx-file.
# raw.data: If FALSE, defined missings are ste to NA and
## variables are convertet to R-classes according to their field type.
## If TRUE, the data.frame contains the original codes as character variables.

# Value
# If the epx-file contains only one data set, the output will be a data.frame.
## It it is a relational data base wit multiple data sets. The outcome will be a list of data frames.
# Study information, variable labels and information about data.frame-relations will be given as attributes.

# This is part of a project to create a package to read epx-files from EpiData into R.

# So far, the functions depend on the package xml2


# Johann Popp
# 2019-06-16
# Last update: 2023-11-27
###########################################

#' Read EpiData epx-Files
#'
#' Reads \href{https://www.epidata.dk/}{EpiData} epx-files into a
#' \code{\link{data.frame}} or a \code{\link{list}} of \code{data.frame}s.
#'
#' @param x An epx-file created by EpiData.
#' @param convert c("all", "none", "missings", "labels", "both") "missings" replaces defined missing values with NA; "labels" replaces value codes with value labels, "both" replaces missing values and value codes and "all" (default) also applies variable classes according to the field types defined in EpiData.
#'
#'
#' @return
#' A simple data base will be returned as a \code{data.frame}. A relational
#' data base will be returned as a \code{list} of \code{data.frame}s.
#'
#' Study informations are stored as \code{attributes(...)$study.info} and variable
#' labels are stored as \code{attributes(...)$variable.labels}. For relational
#' data bases, key variables and parent data sets are stored as \code{attributes(...)$info.relations}.
#'
#' @details
#' By default, object classes of the variables will be set according to the field type defined in EpiData. Coded variables will be returned as \code{factor}s with the value labels defined in EpiData. Values that are defined as missing values in EpiData will be set to \code{NA}.
#'
#' All date and time fields are convertet using *Coordinated Usual Time (UTC)* as time zone argument as EpiData does not provide time zone information by default.
#'
#'
#'
#' @export
#'
#' @examples
#' # Read the example data set "marathon.epx"
#' # (The helper function epx.example is only needed to identify the path
#' # of the example file in the loaded package).
#' dat <- read.EpiData(epx.example("SomeFakeData.epx"))
#' dat
#'
#' # Extract study information
#' attributes(dat)$study.info



read.EpiData <- function(x, convert = "all"){
  if(!any(
    grepl(paste0("^", convert, "$"),
          c("all", "a", "none", "no", "n", "missings", "miss", "m",
            "labels", "label", "l", "both", "b")))){
    stop("Convert can be 'all', 'missings', 'labels', 'both' or 'none'")
  }
  info <- epx.extract(x)
  dat <- lapply(info[[7]], epx.read)

  # Combine dat and info in a data-set-wise list
  perDataSet <- mapply(function(dat, info) list(list(dat = dat, info = info)), dat, info[[7]])

  # Set definded missing variables to NA
  suppressWarnings(
    if(any(grepl(paste0("^", convert, "$"),
                 c("all", "a", "missings", "miss", "m", "both", "b")))){
      dat2 <- lapply(perDataSet, function(x) epx.missing(x$dat, x$info))
      perDataSet <- mapply(function(dat, info){
        list(list(dat = dat, info = info))
      },
      dat2, info[[7]])
    } else {
      dat2 <- dat
    })

  # Apply value labels
  suppressWarnings(
    if(any(
      grepl(paste0("^", convert, "$"),
            c("all", "a", "labels", "label", "l", "both", "b")))){
      dat2 <- lapply(perDataSet,
                     function(x){
                       if(inherits(x[[1]], "data.frame")){
                         epx.labels(x[[1]], x[[2]])
                       } else {"This data frame has no entries"}
                     })
      attributes(dat2) <- attributes(dat)
      perDataSet <- mapply(function(dat, info){
        list(list(dat = dat, info = info))
      },
      dat2, info[[7]])
      dat2
    }
  )

  # Set variable classes
  suppressWarnings(
    if(any(
      grepl(paste0("^", convert, "$"),
            c("all", "a")))){
      dat2 <- lapply(perDataSet,
                     function(x){
                       if(inherits(x[[1]], "data.frame")){
                         epx.class(x[[1]], x[[2]])
                       } else {"This data frame has no entries"}
                     })
      attributes(dat2) <- attributes(dat)
      perDataSet <- mapply(function(dat, info){
        list(list(dat = dat, info = info))
      },
      dat2, info[[7]])
    }
  )


  # Add record status
  dat2 <- lapply(perDataSet,
                 function(x){
                   if(inherits(x[[1]], "data.frame")){
                     data.frame(status = factor(x$info$recordStatus),
                                x$dat)}
                    else{"This data frame has no entries"}})
  attributes(dat2) <- attributes(dat)




  if(length(dat2) == 1){
    dat2 <- dat2[[1]]
  }

  # Extract study information
  infoStudy <- lapply(xml2::xml_children(info$infoStudy), xml2::xml_text)
  names(infoStudy) <- xml2::xml_name(xml2::xml_children(info$infoStudy))
  attributes(dat2)$study.info <- t(data.frame(infoStudy))

  # Collect relational information
  if(length(dat) > 1){
    attributes(dat2)$info.relations <- info$infoRelations
  }

  dat2
}




##############
# The next step will be, to enable the reading of encrypted files
